### ISSUE TITLE
When passing in a K matrix to FoVPerspectiveCameras, the matrix needs to have a specific form, but this restriction does not appear to be documented

### ISSUE THREAD

**Issue by mikeroberts3000** [facebookresearch/pytorch3d#1257]
Hello there, I'm experimenting with supplying a non-standard `K` matrix when constructing an `FoVPerspectiveCameras` object. My interest in this topic is motivated by the non-standard camera intrinsics that were used to render the Hypersim dataset [1].

In my code, I can construct a well-formed `K` matrix from a standard OpenGL perspective projection matrix `M_proj` by left-multiplying an appropriate conversion matrix `M_p3dclip_from_oglclip`, as described in this GitHub issue [2]:

`K = M_p3dclip_from_oglclip * M_proj`

Using this `K` matrix, I can pass it into `FoVPerspectiveCameras` and rendering behaves as expected. For example, here is a Hypersim mesh that I rendered using such a `K` matrix. So far, so good.

![download (12)](https://user-images.githubusercontent.com/2341965/180108650-af6d1718-4ce4-4f37-8acd-9bf94b9d0669.png)

However, If I construct a slightly different `K` matrix by right-multiplying an almost-identity perturbation matrix `M_perturb_cam_pts`, I see major rendering artifacts. For example, suppose I construct `K` as follows:

`K = M_p3dclip_from_oglclip * M_proj * M_perturb_cam_pts`

where

```
M_perturb_cam_pts = np.matrix([[ 1.0,  0.0,   0.0,  0.0 ],
                               [ 0.0,  1.0,   0.0,  0.0 ],
                               [ 0.0,  0.01,  1.0,  0.0 ],
                               [ 0.0,  0.0,   0.0,  1.0 ]])
```

In this case, here is the rendering output I get.

![download (13)](https://user-images.githubusercontent.com/2341965/180108664-688d6c77-90ab-437b-b2b3-4208c7cf5b03.png)

I don't understand this result. If I right-multiply my working `K` matrix by `M_perturb_cam_pts`, this should be equivalent to applying a small perturbation to all my vertices in camera-space, and should not cause such major artifacts. Indeed, I've tried transforming all my vertices into camera-space in my own code (including applying `M_perturb_cam_pts`) and then rendering with an identity `FoVPerspectiveCameras`, and rendering behaves as expected. I see an ever-so-slightly perturbed image, which makes sense because all the vertices have been slightly perturbed in camera-space. I have also implemented these camera-space perturbations in an OpenGL reference renderer, and again rendering behaves as expected. But somehow, encoding `M_perturb_cam_pts` directly into my `K` matrix and passing it into `FoVPerspectiveCameras` causes major rendering artifacts in PyTorch3D. It appears that clipping doesn't work properly if the user supplies a `K` matrix with non-standard transformations encoded into it. This naturally leads to the question, what exactly is allowed and is not allowed in the user-supplied `K` matrix?

In summary, there appear to be some implicit restrictions on the form of the user-supplied `K` matrix, but these restrictions are not clearly documented. It would be helpful if there was an extra sentence or two in the `FoVPerspectiveCameras` documentation describing the required form of `K`.

I'll post a complete code snippet in a follow-up post.

[1] https://github.com/apple/ml-hypersim
[2] https://github.com/facebookresearch/pytorch3d/issues/1255

**Comment by gkioxari** [comment]
HI @mikeroberts3000 
I had a very similar experience with a user-defined `K` matrix with `FoVPerspectiveCameras`. I will look into this more closely when I come back from my trip at the end of the week because I want to resolve this asap!

**Comment by gkioxari** [comment]
@mikeroberts3000 Can you send me the vertices and faces of the scene in your code above? Email me! :-D

I am working with a different scene, `ai_037_002`, and I can't reproduce this. For `ai_037_002`, I get the following renderings with non-perturbed (top) and perturbed (bottom) projection matrix.

![image](https://user-images.githubusercontent.com/4369065/183252601-d5b443b1-595d-400b-b7de-7237e5d6a985.png)
![image](https://user-images.githubusercontent.com/4369065/183252616-285715b1-f7da-4df1-aff3-e6c412018c75.png)

**Comment by mikeroberts3000** [comment]
Ah, that is a super interesting result! I think somehow the bug depends on the magnitude of the perturbation matrix and the scene geometry (but I have no idea how). In other words, maybe the perturbation matrix in my code is enough to cause problems for the scene I'm using (ai_001_001), but not the scene you're using (ai_037_002). If you increase the non-zero off-diagonal entry from 0.01 to 0.05, do you start to see problems?

I'll follow up over email to send you the mesh data!

**Comment by gkioxari** [comment]
Aha! So this is not a bug. It all lies in PyTorch3D's `zbuf` definition during rendering.  Let me explain (verbosely!)

## [Background] The sequence of transforms in PyTorch3d rendering

Assume `points_world` are the input points (world), in our case vertices of a scene mesh. The input points undergo two transforms
* `world ->  camera`: transforms `points_world` to `points_cam` with world-to-camera transforms `(R, T)`
* `camera -> ndc`: projects `points_cam` to `points_ndc` with camera matrix `K`

In rasterization, we use z-buffering. But which z? In the case of PyTorch3D z-buffering stores the depth of `points_cam`, in other words the depth of points *before the camera projection `K` is applied*.

https://github.com/facebookresearch/pytorch3d/blob/276c9a8acbfa715f5802f26ec9f4141bde26ecb4/pytorch3d/renderer/mesh/rasterizer.py#L207
(in the code `view` is the same as `camera` in the explanation above)

## The impact of perturbation for `zbuf`

Now, Mike is adding another transform, namely a perturbation
* `camera -> perturbed`: perturbs `points_cam` to `points_cam_pert` with `M_perturb`

This perturbation migth alter the depth of `points_cam` (e.g. in Mike's perturbation `z_cam_pert = 0.01 * y_cam + 1.0 * z_cam`). **Well, now we are in a pickle! So what should the renderer use for z-buffering, the perturbed `z_cam_pert` or the original `z_cam` ??**

The way Mike has set up the rendering above, PyTorch3D will use the original `z_cam` in the z-buffer but will render `points_cam_pert`. This is because the perturbation is defined as part of the camera projection matrix. In other words, points will be ordered based on `z_cam` but the points rendered will be `points_cam_pert != points_cam`. 

In Mike's first `(R, T)` transform, `points_cam ~= points_cam_pert` so the rendering looks similar-ish. But if `points_cam != points_cam_pert` (as in Mike's second `(R,T)` transform), things will be off.  

## The right way to perturb with PyTorch3D

Because of PyTorch3D's choice to use as z-buffer the depth of points before the camera projection is applied, we have to re-think how to implement a perturbation. I believe that right solution is for z-buffering to use `z_cam_pert` when rendering `points_cam_pert`. This means that we can't define perturbation as part of the camera matrix. Instead do this:

``` python

# manually perform world -> camera transform
R, T = look_at_view_transform(eye=[[90.0, -90.0, 90.0]], up=[[0.0, 0.0, 1.0]], at=[[0.0, 0.0, 0.0]])
vertices_rot = torch.matmul(vertices, R[0].to(device)) + T[0].to(device) # R, T here are column-order

# perturb camera points
K_pert = torch.tensor(M_perturb_cam_pts[:3, :3], device=device, dtype=torch.float32)
vertices_rot_pert = torch.matmul(vertices_rot, K_pert.transpose(0, 1))
mesh_rot_pert = Meshes(verts=[vertices_rot_pert], faces=[faces])

# define a camera without incorporating perturbation (we have perturbed the points above)
camera = FoVPerspectiveCameras(device=device, K=[M_p3dclip_from_clip @ M_proj])

raster_settings = RasterizationSettings(
    image_size=[height_pixels, width_pixels],
    blur_radius=0.0,
    faces_per_pixel=1,
)

rasterizer = MeshRasterizer(cameras=camera, raster_settings=raster_settings)
renderer = MeshRenderer(
    rasterizer=rasterizer,
    shader=HardDepthShader(device=device, cameras=camera),
)
```

The output now is fixed. 
Below we show Mike's unperturbed output, Mike's perturbed output and our perturbed output

Mike's unperturbed
![image](https://user-images.githubusercontent.com/4369065/183898282-ee13359c-e386-497c-ac1f-21344d6fdc54.png)
Mike's perturbed (which was wrong because of the z-buffer discrepancy)
![image](https://user-images.githubusercontent.com/4369065/183898307-5ab8085c-7d77-4cd1-bd45-f39e9992cee0.png)
Our perturbed 
![image](https://user-images.githubusercontent.com/4369065/183898385-e90dc144-65e0-4fd8-b327-ee8cf33102e4.png)

**Comment by mikeroberts3000** [comment]
Ahh thank you so much @gkioxari. Your explanation makes perfect sense, and now we know exactly what kinds of `K` matrices are allowed in PyTorch3D (they're not allowed to change the z-order of the vertices). And your explanation motivates the workaround that has been working for both of us, i.e., to apply camera-space perturbations manually outside of `FovPerspectiveCameras`. Yay!
